### DISCUSSION TITLE
OK-Formatted Test Cells vs. Exception-Based Test Cells

### DISCUSSION THREAD

**Discussion by vladtkachuk4** [ucbds-infra/otter-grader#864] [Q&A] [answered]
I'm trying to decide between using the OK-formatted test cells and exception-based test cells.

It seems that the OK-formatted test cells are quicker to write, don't require you to pass in necessary variables as function arguments, and what I like the most is that I don't have to specify a new function name for each test.

I'm wondering if there are any negatives or things I should be aware of when choosing to use OK-formatted tests instead of exception-based ones? For instance, is there something I can't do with OK-formatted that I can do with exception-based?

**Comment by chrispyles** [comment, marked as answer]
There are two main things to note about OK tests:

* only the last line of a test cell may produce output; if other lines do things like print values, they must be ended with a semicolon
* OK tests use doctests under the hood, so values are compared by comparing their `repr`s

You can find a brief guide on writing autograder tests, focusing on OK tests, [here](https://autograder-tests.readthedocs.io/en/latest/). It describes some of the common pitfalls of OK tests.

Besides these two things and the limitations of exception-based tests you mentioned, they both have the same capabilities. As long as you keep in mind the items above, I personally find OK tests more ergonomic.

**Reply by vladtkachuk4** [reply]
Very helpful response, thank you!
